In [ ]:
import numpy as np
import pandas as pd
import html5lib

In [ ]:
titanic = pd.read_html('https://github.com/ogrisel/parallel_ml_tutorial/blob/master/notebooks/titanic_train.csv', header = 0)
titanic = titanic[0]
titanic = titanic.drop(titanic.columns[0], axis=1) # уберемнеинформативнуюколонку
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
titanic.describe()


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [ ]:
titanic[(titanic["Embarked"] == "C") & (titanic.Fare > 200)].head()

PassengerId  Survived  Pclass  \
118          119         0       1   
258          259         1       1   
299          300         1       1   
311          312         1       1   
377          378         0       1   

                                                Name     Sex   Age  SibSp  \
118                         Baxter, Mr. Quigg Edmond    male  24.0      0   
258                                 Ward, Miss. Anna  female  35.0      0   
299  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                       Ryerson, Miss. Emily Borie  female  18.0      2   
377                        Widener, Mr. Harry Elkins    male  27.0      0   

     Parch    Ticket      Fare            Cabin Embarked  
118      1  PC 17558  247.5208          B58 B60        C  
258      0  PC 17755  512.3292              NaN        C  
299      1  PC 17558  247.5208          B58 B60        C  
311      2  PC 17608  262.3750  B57 B59 B63 B66        C  
377      2    113503  211.5000              C82        C

In [ ]:
titanic[(titanic["Embarked"] == "C") & (titanic["Fare"] > 200)].sort_values(by="Fare", ascending=False).head()


PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                 Lesurer, Mr. Gustave J   
311          312         1       1             Ryerson, Miss. Emily Borie   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

In [ ]:
def age_category(age):
    """
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    """
    if age < 30:
      return 1
    elif age < 55:
        return 2
    else:
        return 3

In [ ]:
age_categories = [age_category(age) for age in titanic.Age]
titanic["Age_category"] = age_categories

In [ ]:
titanic["Age_category"] = titanic["Age"].apply(age_category)


1. Сколько мужчин / женщин находилось на борту?


In [ ]:
titanic['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

На борту находилось 577 мужчин и 314 женщин

2. Выведите распределение переменной Pclass (социально-экономический статус) и это же распределение, только для мужчин /
женщин по отдельности. Сколько было мужчин 2-го класса?


In [ ]:
titanic['Pclass'].value_counts()


3    491
1    216
2    184
Name: Pclass, dtype: int64

In [ ]:
titanic[(titanic['Pclass']) & (titanic['Sex'] == 'male')]['Pclass'].value_counts()

3    347
1    122
Name: Pclass, dtype: int64

Почему-то не находит 2 класс, но если задать явно, то выводится

In [ ]:
titanic[(titanic['Pclass'] == 2) & (titanic['Sex'] == 'male')]['Pclass'].value_counts()

2    108
Name: Pclass, dtype: int64

In [ ]:
titanic[(titanic['Pclass']) & (titanic['Sex'] == 'female')]['Pclass'].value_counts()

3    144
1     94
Name: Pclass, dtype: int64

In [ ]:
titanic[(titanic['Pclass']==2) & (titanic['Sex'] == 'female')]['Pclass'].value_counts()


2    76
Name: Pclass, dtype: int64

Мужчин 1 класса-122, 2 класса - 108, 3 класса -347
Женщин 1 класса - 94, 2 класса-76, 3 класса - 144

3. Каковы медиана и стандартное отклонение платежей (Fare)? Округлите до 2 десятичных знаков.

In [ ]:
round(titanic['Fare'].std(), 2)

49.69

In [ ]:
round(titanic['Fare'].median(), 2)

14.45

стандартное отклонение платежей (Fare) - 49.49,
Медиана - 14.45

4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?

In [ ]:
young = titanic[titanic['Age'] < 30]['Survived']
old = titanic[titanic['Age'] > 60]['Survived']
print(f'young { round(100 * young.mean(), 1)}%, old { round(100 * old.mean(), 1)}%')

young 40.6%, old 22.7%


5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?


In [ ]:
male = titanic[titanic['Sex'] == 'male']['Survived']
female = titanic[titanic['Sex'] =='female']['Survived']
print(f'Выжившие среди мужчин { round(100 * male.mean(), 1)}%, женщин { round(100 * female.mean(), 1)}%')

Выжившие среди мужчин 18.9%, женщин 74.2%


6. Найдите самое популярное имя среди пассажиров Титаника мужского пола?

In [ ]:
def first_name(name):
  return name.split(',')[1].split()[1]

In [ ]:
popular_name = titanic[titanic['Sex'] == 'male']['Name'].apply(first_name)
popular_name.value_counts().head()

William    35
John       25
George     14
Charles    13
Thomas     13
Name: Name, dtype: int64

7. Сравните распределение возраста у спасенных и у погибших. Средний возраст погибших выше, верно?

In [ ]:
titanic[titanic['Survived'] == 1]['Age'].value_counts()

24.00    15
36.00    11
35.00    11
27.00    11
22.00    11
         ..
55.00     1
32.50     1
12.00     1
0.92      1
47.00     1
Name: Age, Length: 65, dtype: int64

In [ ]:
titanic[titanic['Survived'] == 0]['Age'].value_counts()

21.0    19
28.0    18
25.0    17
18.0    17
22.0    16
        ..
15.0     1
36.5     1
23.5     1
55.5     1
74.0     1
Name: Age, Length: 77, dtype: int64

In [ ]:
titanic.groupby('Survived')['Age'].mean()

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64

8. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:


In [ ]:
pd.crosstab(titanic['Pclass'], titanic['Sex'], 
            values=titanic['Age'], aggfunc=np.mean)

Sex        female       male
Pclass                      
1       34.611765  41.281386
2       28.722973  30.740707
3       21.750000  26.507589